File: source/modulo_components/test/cpp/test_component_communication.cpp
#include <gtest/gtest.h>

#include "modulo_components/Component.hpp"
#include "test_modulo_components/communication_components.hpp"

using namespace modulo_components;

class ComponentCommunicationTest : public ::testing::Test {
protected:
  void SetUp() override {
    rclcpp::init(0, nullptr);
    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
  }

  void TearDown() override {
    rclcpp::shutdown();
  }

  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
};

TEST_F(ComponentCommunicationTest, InputOutput) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<MinimalInput<Component>>(
      rclcpp::NodeOptions().parameter_overrides({rclcpp::Parameter("period", 0.1)}), "/topic"
  );
  auto output_node = std::make_shared<MinimalOutput<Component>>(
      rclcpp::NodeOptions().parameter_overrides({rclcpp::Parameter("period", 0.1)}), "/topic", cartesian_state
  );
  this->exec_->add_node(input_node);
  this->exec_->add_node(output_node);
  this->exec_->template spin_until_future_complete(input_node->received_future, 500ms);
  EXPECT_EQ(cartesian_state.get_name(), input_node->input->get_name());
  EXPECT_TRUE(cartesian_state.data().isApprox(input_node->input->data()));
  this->exec_.reset();
}

File: source/modulo_components/test/python/test_component_communication.py
import pytest
from modulo_components.component import Component


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[Component, "/topic"]], indirect=True)
def test_input_output(ros_exec, random_state, minimal_cartesian_output, minimal_cartesian_input):
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    ros_exec.remove_node(minimal_cartesian_input)
    ros_exec.remove_node(minimal_cartesian_output)
    assert minimal_cartesian_input.received_future.result()
    assert random_state.get_name() == minimal_cartesian_input.input.get_name()
    assert random_state.dist(minimal_cartesian_input.input) < 1e-3


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
def test_input_output_invalid(ros_exec, make_minimal_invalid_encoded_state_publisher, minimal_cartesian_input):
    invalid_publisher = make_minimal_invalid_encoded_state_publisher("/topic")
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(invalid_publisher)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.result()
